File: Excel/542 Challenge.py
import pandas as pd
import numpy as np

path = "542 Squares from Strings.xlsx"
test_abc = pd.read_excel(path, sheet_name=None, usecols="C:E", skiprows=1, nrows=3, header=None).get('Sheet1').replace(np.nan, '', regex=True)
test_abcd = pd.read_excel(path, sheet_name=None, usecols="C:F", skiprows=5, nrows=4, header=None).get('Sheet1').replace(np.nan, '', regex=True)
test_microsoft = pd.read_excel(path, sheet_name=None, usecols="C:K", skiprows=10, nrows=9, header=None).get('Sheet1').replace(np.nan, '', regex=True)

def make_word_frame(word):
    n = len(word)
    word_frame = np.full((n, n), '', dtype=str)
    word_frame[0] = list(word)
    word_frame[:, 0] = list(word)
    word_frame[-1] = list(word[::-1])
    word_frame[:, -1] = list(word[::-1])
    return word_frame

print(all(make_word_frame("abc")==test_abc)) # True
print(all(make_word_frame("abcd")==test_abcd)) # True
print(all(make_word_frame("microsoft")==test_microsoft)) # True

File: Excel/500-599/542/542 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/542 Squares from Strings.xlsx"
test_abc = read_excel(path, range = "C2:E4", col_names = FALSE) %>% as.matrix()
test_abcd = read_excel(path, range = "C6:F9", col_names = FALSE) %>% as.matrix()
test_microsoft = read_excel(path, range = "C11:K19", col_names = FALSE) %>% as.matrix()

make_word_frame <- function(word) {
  n <- nchar(word)
  chars <- str_split(word, "")[[1]]
  M <- matrix(NA, n, n)
  
  M[1, ] <- M[, 1] <- chars
  M[n, ] <- M[, n] <- rev(chars)
  
  return(M)
}

all.equal(make_word_frame("abc"), test_abc, check.attributes = FALSE) # TRUE
all.equal(make_word_frame("abcd"), test_abcd, check.attributes = FALSE) # TRUE
all.equal(make_word_frame("microsoft"), test_microsoft, check.attributes = FALSE) # TRUE

